Caching not working due to the new channel type

Due to the new "Radio" channel type, if there is a channel created with the Radio type, Javacord will no longer be able to cache the channel and will throw an `AssertionError` in `ServerImpl` at line 379.

```
java.lang.AssertionError: null
	at java.util.Optional.orElseThrow(Optional.java:401) ~[?:?]
	at org.javacord.core.entity.server.ServerImpl.<init>(ServerImpl.java:379) ~[javacord-core-3.1.2.jar:?]
	at org.javacord.core.util.handler.guild.GuildCreateHandler.handle(GuildCreateHandler.java:33) ~[javacord-core-3.1.2.jar:?]
	at org.javacord.core.util.gateway.PacketHandler.lambda$handlePacket$0(PacketHandler.java:51) ~[javacord-core-3.1.2.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
	at java.lang.Thread.run(Thread.java:832) [?:?]
```

--- a/javacord-api/src/main/java/org/javacord/api/DiscordApi.java
+++ b/javacord-api/src/main/java/org/javacord/api/DiscordApi.java
@@ -9,6 +9,7 @@ import org.javacord.api.entity.channel.ChannelCategory;
 import org.javacord.api.entity.channel.GroupChannel;
 import org.javacord.api.entity.channel.PrivateChannel;
 import org.javacord.api.entity.channel.ServerChannel;
+import org.javacord.api.entity.channel.ServerStageVoiceChannel;
 import org.javacord.api.entity.channel.ServerTextChannel;
 import org.javacord.api.entity.channel.ServerVoiceChannel;
 import org.javacord.api.entity.channel.TextChannel;
@@ -1244,6 +1245,13 @@ public interface DiscordApi extends GloballyAttachableListenerManager {
     Collection<ServerVoiceChannel> getServerVoiceChannels();
 
     /**
+     * Gets a collection with all server stage voice channels of the bot.
+     *
+     * @return A collection with all server stage voice channels of the bot.
+     */
+    Collection<ServerStageVoiceChannel> getServerStageVoiceChannels();
+
+    /**
      * Gets a collection with all text channels of the bot.
      *
      * @return A collection with all text channels of the bot.
@@ -1615,6 +1623,58 @@ public interface DiscordApi extends GloballyAttachableListenerManager {
     default Collection<ServerVoiceChannel> getServerVoiceChannelsByNameIgnoreCase(String name) {
         return Collections.unmodifiableList(
                 getServerVoiceChannels().stream()
+                        .filter(channel -> channel.getName().equalsIgnoreCase(name))
+                        .collect(Collectors.toList()));
+    }
+
+    /**
+     * Gets a server stage voice channel by its id.
+     *
+     * @param id The id of the server stage voice channel.
+     * @return The server stage voice channel with the given id.
+     */
+    default Optional<ServerStageVoiceChannel> getServerStageVoiceChannelById(long id) {
+        return getChannelById(id).flatMap(Channel::asServerStageVoiceChannel);
+    }
+
+    /**
+     * Gets a server stage voice channel by its id.
+     *
+     * @param id The id of the server stage voice channel.
+     * @return The server stage voice channel with the given id.
+     */
+    default Optional<ServerStageVoiceChannel> getServerStageVoiceChannelById(String id) {
+        try {
+            return getServerStageVoiceChannelById(Long.parseLong(id));
+        } catch (NumberFormatException e) {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Gets a collection with all server stage voice channels with the given name.
+     * This method is case sensitive!
+     *
+     * @param name The name of the server stage voice channels.
+     * @return A collection with all server stage voice channels with the given name.
+     */
+    default Collection<ServerStageVoiceChannel> getServerStageVoiceChannelsByName(String name) {
+        return Collections.unmodifiableList(
+                getServerStageVoiceChannels().stream()
+                        .filter(channel -> channel.getName().equals(name))
+                        .collect(Collectors.toList()));
+    }
+
+    /**
+     * Gets a collection with all server stage voice channels with the given name.
+     * This method is case insensitive!
+     *
+     * @param name The name of the server stage voice channels.
+     * @return A collection with all server stage voice channels with the given name.
+     */
+    default Collection<ServerStageVoiceChannel> getServerStageVoiceChannelsByNameIgnoreCase(String name) {
+        return Collections.unmodifiableList(
+                getServerStageVoiceChannels().stream()
                         .filter(channel -> channel.getName().equalsIgnoreCase(name))
                         .collect(Collectors.toList()));
     }

--- a/javacord-api/src/main/java/org/javacord/api/entity/channel/Channel.java
+++ b/javacord-api/src/main/java/org/javacord/api/entity/channel/Channel.java
@@ -88,6 +88,15 @@ public interface Channel extends DiscordEntity, UpdatableFromCache, ChannelAttac
     }
 
     /**
+     * Gets the channel as server stage voice channel.
+     *
+     * @return The channel as server stage voice channel.
+     */
+    default Optional<ServerStageVoiceChannel> asServerStageVoiceChannel() {
+        return as(ServerStageVoiceChannel.class);
+    }
+
+    /**
      * Gets the channel as text channel.
      *
      * @return The channel as text channel.

--- a/javacord-api/src/main/java/org/javacord/api/entity/channel/ChannelType.java
+++ b/javacord-api/src/main/java/org/javacord/api/entity/channel/ChannelType.java
@@ -14,6 +14,7 @@ public enum ChannelType {
     CHANNEL_CATEGORY(4, false, false, true),
     SERVER_NEWS_CHANNEL(5, true, false, true),
     SERVER_STORE_CHANNEL(6, true, false, true),
+    SERVER_STAGE_VOICE_CHANNEL(13, false, true, true),
     UNKNOWN(-1, false, false, false);
 
     private static final ChannelType[] textChannelTypes = Arrays.stream(ChannelType.values())

--- a/javacord-api/src/main/java/org/javacord/api/entity/channel/ServerStageVoiceChannel.java
+++ b/javacord-api/src/main/java/org/javacord/api/entity/channel/ServerStageVoiceChannel.java
@@ -1,0 +1,15 @@
+package org.javacord.api.entity.channel;
+
+import org.javacord.api.listener.channel.server.voice.ServerStageVoiceChannelAttachableListenerManager;
+
+import java.util.Optional;
+
+public interface ServerStageVoiceChannel extends ServerVoiceChannel, ServerStageVoiceChannelAttachableListenerManager {
+
+    /**
+     * Gets the topic of this.
+     *
+     * @return The topic of this channel.
+     */
+    Optional<String> getTopic();
+}

--- a/javacord-api/src/main/java/org/javacord/api/entity/channel/ServerStageVoiceChannel.java
+++ b/javacord-api/src/main/java/org/javacord/api/entity/channel/ServerStageVoiceChannel.java
@@ -6,6 +6,11 @@ import java.util.Optional;
 
 public interface ServerStageVoiceChannel extends ServerVoiceChannel, ServerStageVoiceChannelAttachableListenerManager {
 
+    @Override
+    default ChannelType getType() {
+        return ChannelType.SERVER_STAGE_VOICE_CHANNEL;
+    }
+
     /**
      * Gets the topic of this.
      *

--- a/javacord-api/src/main/java/org/javacord/api/entity/server/Server.java
+++ b/javacord-api/src/main/java/org/javacord/api/entity/server/Server.java
@@ -13,6 +13,7 @@ import org.javacord.api.entity.auditlog.AuditLogEntry;
 import org.javacord.api.entity.channel.ChannelCategory;
 import org.javacord.api.entity.channel.ChannelCategoryBuilder;
 import org.javacord.api.entity.channel.ServerChannel;
+import org.javacord.api.entity.channel.ServerStageVoiceChannel;
 import org.javacord.api.entity.channel.ServerTextChannel;
 import org.javacord.api.entity.channel.ServerTextChannelBuilder;
 import org.javacord.api.entity.channel.ServerVoiceChannel;
@@ -2349,6 +2350,32 @@ public interface Server extends DiscordEntity, Nameable, UpdatableFromCache<Serv
     default Optional<ServerVoiceChannel> getVoiceChannelById(String id) {
         try {
             return getVoiceChannelById(Long.parseLong(id));
+        } catch (NumberFormatException e) {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Gets a stage voice channel by its id.
+     *
+     * @param id The id of the stage voice channel.
+     * @return The stage voice channel with the given id.
+     */
+    default Optional<ServerStageVoiceChannel> getStageVoiceChannelById(long id) {
+        return getChannelById(id)
+                .filter(channel -> channel instanceof ServerStageVoiceChannel)
+                .map(channel -> (ServerStageVoiceChannel) channel);
+    }
+
+    /**
+     * Gets a voice channel by its id.
+     *
+     * @param id The id of the voice channel.
+     * @return The voice channel with the given id.
+     */
+    default Optional<ServerStageVoiceChannel> getStageVoiceChannelById(String id) {
+        try {
+            return getStageVoiceChannelById(Long.parseLong(id));
         } catch (NumberFormatException e) {
             return Optional.empty();
         }

--- a/javacord-api/src/main/java/org/javacord/api/event/channel/server/voice/ServerStageVoiceChannelChangeTopicEvent.java
+++ b/javacord-api/src/main/java/org/javacord/api/event/channel/server/voice/ServerStageVoiceChannelChangeTopicEvent.java
@@ -1,0 +1,23 @@
+package org.javacord.api.event.channel.server.voice;
+
+import org.javacord.api.entity.channel.ServerStageVoiceChannel;
+
+public interface ServerStageVoiceChannelChangeTopicEvent extends ServerVoiceChannelEvent {
+
+    @Override
+    ServerStageVoiceChannel getChannel();
+
+    /**
+     * Gets the new topic of the channel.
+     *
+     * @return The new topic of the channel.
+     */
+    String getNewTopic();
+
+    /**
+     * Gets the old topic of the channel.
+     *
+     * @return The old topic of the channel.
+     */
+    String getOldTopic();
+}

--- a/javacord-api/src/main/java/org/javacord/api/listener/channel/server/voice/ServerStageVoiceChannelAttachableListener.java
+++ b/javacord-api/src/main/java/org/javacord/api/listener/channel/server/voice/ServerStageVoiceChannelAttachableListener.java
@@ -1,0 +1,10 @@
+package org.javacord.api.listener.channel.server.voice;
+
+import org.javacord.api.entity.channel.ServerStageVoiceChannel;
+
+/**
+ * This is a marker interface for listeners that can be attached to a {@link ServerStageVoiceChannel}.
+ */
+public interface ServerStageVoiceChannelAttachableListener {
+
+}

--- a/javacord-api/src/main/java/org/javacord/api/listener/channel/server/voice/ServerStageVoiceChannelChangeTopicListener.java
+++ b/javacord-api/src/main/java/org/javacord/api/listener/channel/server/voice/ServerStageVoiceChannelChangeTopicListener.java
@@ -1,0 +1,18 @@
+package org.javacord.api.listener.channel.server.voice;
+
+import org.javacord.api.event.channel.server.voice.ServerStageVoiceChannelChangeTopicEvent;
+import org.javacord.api.listener.GloballyAttachableListener;
+import org.javacord.api.listener.ObjectAttachableListener;
+import org.javacord.api.listener.server.ServerAttachableListener;
+
+public interface ServerStageVoiceChannelChangeTopicListener  extends ServerAttachableListener,
+        ServerStageVoiceChannelAttachableListener,
+        GloballyAttachableListener, ObjectAttachableListener {
+
+    /**
+     * This method is called every time a server stage voice channel's topic changes.
+     *
+     * @param event The event.
+     */
+    void onServerStageVoiceChannelChangeTopic(ServerStageVoiceChannelChangeTopicEvent event);
+}

--- a/javacord-core/src/main/java/org/javacord/core/DiscordApiImpl.java
+++ b/javacord-core/src/main/java/org/javacord/core/DiscordApiImpl.java
@@ -19,6 +19,7 @@ import org.javacord.api.entity.channel.ChannelType;
 import org.javacord.api.entity.channel.GroupChannel;
 import org.javacord.api.entity.channel.PrivateChannel;
 import org.javacord.api.entity.channel.ServerChannel;
+import org.javacord.api.entity.channel.ServerStageVoiceChannel;
 import org.javacord.api.entity.channel.ServerTextChannel;
 import org.javacord.api.entity.channel.ServerVoiceChannel;
 import org.javacord.api.entity.channel.TextChannel;
@@ -1763,6 +1764,11 @@ public class DiscordApiImpl implements DiscordApi, DispatchQueueSelector {
     @Override
     public Collection<ServerVoiceChannel> getServerVoiceChannels() {
         return entityCache.get().getChannelCache().getChannelsWithTypes(ChannelType.SERVER_VOICE_CHANNEL);
+    }
+
+    @Override
+    public Collection<ServerStageVoiceChannel> getServerStageVoiceChannels() {
+        return entityCache.get().getChannelCache().getChannelsWithTypes(ChannelType.SERVER_STAGE_VOICE_CHANNEL);
     }
 
     @Override

--- a/javacord-core/src/main/java/org/javacord/core/entity/channel/ServerStageVoiceChannelImpl.java
+++ b/javacord-core/src/main/java/org/javacord/core/entity/channel/ServerStageVoiceChannelImpl.java
@@ -1,0 +1,51 @@
+package org.javacord.core.entity.channel;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import org.javacord.api.entity.channel.ServerStageVoiceChannel;
+import org.javacord.core.DiscordApiImpl;
+import org.javacord.core.entity.server.ServerImpl;
+import org.javacord.core.listener.channel.server.voice.InternalServerStageVoiceChannelAttachableListenerManager;
+
+import java.util.Optional;
+
+public class ServerStageVoiceChannelImpl extends ServerVoiceChannelImpl
+        implements ServerStageVoiceChannel, InternalServerStageVoiceChannelAttachableListenerManager {
+
+    /**
+     * The topic of the stage.
+     */
+    private String topic;
+
+    /**
+     * Creates a new server stage voice channel object.
+     *
+     * @param api The discord api instance.
+     * @param server The server of the channel.
+     * @param data The json data of the channel.
+     */
+    public ServerStageVoiceChannelImpl(DiscordApiImpl api, ServerImpl server, JsonNode data) {
+        super(api, server, data);
+
+        if (data.hasNonNull("topic")) {
+            topic = data.get("topic").asText();
+        } else {
+            topic = null;
+        }
+
+    }
+
+    @Override
+    public Optional<String> getTopic() {
+        return Optional.ofNullable(topic);
+    }
+
+    /**
+     * Sets the topic of the channel.
+     *
+     * @param topic The new topic of the channel.
+     */
+    public void setTopic(String topic) {
+        this.topic = topic;
+    }
+
+}

--- a/javacord-core/src/main/java/org/javacord/core/entity/channel/ServerStageVoiceChannelImpl.java
+++ b/javacord-core/src/main/java/org/javacord/core/entity/channel/ServerStageVoiceChannelImpl.java
@@ -48,4 +48,9 @@ public class ServerStageVoiceChannelImpl extends ServerVoiceChannelImpl
         this.topic = topic;
     }
 
+    @Override
+    public String toString() {
+        return String.format("ServerStageVoiceChannel (id: %s, name: %s)", getIdAsString(), getName());
+    }
+
 }

--- a/javacord-core/src/main/java/org/javacord/core/entity/server/ServerImpl.java
+++ b/javacord-core/src/main/java/org/javacord/core/entity/server/ServerImpl.java
@@ -19,6 +19,7 @@ import org.javacord.api.entity.auditlog.AuditLogEntry;
 import org.javacord.api.entity.channel.ChannelCategory;
 import org.javacord.api.entity.channel.ChannelType;
 import org.javacord.api.entity.channel.ServerChannel;
+import org.javacord.api.entity.channel.ServerStageVoiceChannel;
 import org.javacord.api.entity.channel.ServerTextChannel;
 import org.javacord.api.entity.channel.ServerVoiceChannel;
 import org.javacord.api.entity.emoji.KnownCustomEmoji;
@@ -46,6 +47,7 @@ import org.javacord.core.entity.activity.ActivityImpl;
 import org.javacord.core.entity.auditlog.AuditLogImpl;
 import org.javacord.core.entity.channel.ChannelCategoryImpl;
 import org.javacord.core.entity.channel.ServerChannelImpl;
+import org.javacord.core.entity.channel.ServerStageVoiceChannelImpl;
 import org.javacord.core.entity.channel.ServerTextChannelImpl;
 import org.javacord.core.entity.channel.ServerVoiceChannelImpl;
 import org.javacord.core.entity.permission.RoleImpl;
@@ -337,12 +339,17 @@ public class ServerImpl implements Server, Cleanupable, InternalServerAttachable
 
         if (data.has("channels")) {
             for (JsonNode channel : data.get("channels")) {
+
                 switch (ChannelType.fromId(channel.get("type").asInt())) {
                     case SERVER_TEXT_CHANNEL:
                         getOrCreateServerTextChannel(channel);
                         break;
                     case SERVER_VOICE_CHANNEL:
                         getOrCreateServerVoiceChannel(channel);
+                        break;
+                    case SERVER_STAGE_VOICE_CHANNEL:
+                        // TODO Handle them different from voice channel:
+                        getOrCreateServerStageVoiceChannel(channel);
                         break;
                     case CHANNEL_CATEGORY:
                         getOrCreateChannelCategory(channel);
@@ -717,6 +724,24 @@ public class ServerImpl implements Server, Cleanupable, InternalServerAttachable
         synchronized (this) {
             if (type == ChannelType.SERVER_VOICE_CHANNEL) {
                 return getVoiceChannelById(id).orElseGet(() -> new ServerVoiceChannelImpl(api, this, data));
+            }
+        }
+        // Invalid channel type
+        return null;
+    }
+
+    /**
+     * Gets or creates a server stage voice channel.
+     *
+     * @param data The json data of the channel.
+     * @return The server stage voice channel.
+     */
+    public ServerStageVoiceChannel getOrCreateServerStageVoiceChannel(JsonNode data) {
+        long id = Long.parseLong(data.get("id").asText());
+        ChannelType type = ChannelType.fromId(data.get("type").asInt());
+        synchronized (this) {
+            if (type == ChannelType.SERVER_STAGE_VOICE_CHANNEL) {
+                return getStageVoiceChannelById(id).orElseGet(() -> new ServerStageVoiceChannelImpl(api, this, data));
             }
         }
         // Invalid channel type

--- a/javacord-core/src/main/java/org/javacord/core/entity/server/ServerImpl.java
+++ b/javacord-core/src/main/java/org/javacord/core/entity/server/ServerImpl.java
@@ -348,7 +348,6 @@ public class ServerImpl implements Server, Cleanupable, InternalServerAttachable
                         getOrCreateServerVoiceChannel(channel);
                         break;
                     case SERVER_STAGE_VOICE_CHANNEL:
-                        // TODO Handle them different from voice channel:
                         getOrCreateServerStageVoiceChannel(channel);
                         break;
                     case CHANNEL_CATEGORY:

--- a/javacord-core/src/main/java/org/javacord/core/event/channel/server/voice/ServerStageVoiceChannelChangeTopicEventImpl.java
+++ b/javacord-core/src/main/java/org/javacord/core/event/channel/server/voice/ServerStageVoiceChannelChangeTopicEventImpl.java
@@ -1,0 +1,49 @@
+package org.javacord.core.event.channel.server.voice;
+
+import org.javacord.api.entity.channel.ServerStageVoiceChannel;
+import org.javacord.api.event.channel.server.voice.ServerStageVoiceChannelChangeTopicEvent;
+
+public class ServerStageVoiceChannelChangeTopicEventImpl extends ServerVoiceChannelEventImpl
+        implements ServerStageVoiceChannelChangeTopicEvent {
+
+    /**
+     * The new topic of the channel.
+     */
+    private final String newTopic;
+
+    /**
+     * The old topic of the channel.
+     */
+    private final String oldTopic;
+
+    /**
+     * Creates a new server stage voice channel change topic event.
+     *
+     * @param channel The server stage voice channel of the event.
+     * @param newTopic The new topic of the channel.
+     * @param oldTopic The old topic of the channel.
+     */
+
+    public ServerStageVoiceChannelChangeTopicEventImpl(ServerStageVoiceChannel channel, String newTopic,
+                                                       String oldTopic) {
+        super(channel);
+        this.newTopic = newTopic;
+        this.oldTopic = oldTopic;
+    }
+
+    @Override
+    public String getNewTopic() {
+        return newTopic;
+    }
+
+    @Override
+    public String getOldTopic() {
+        return oldTopic;
+    }
+
+    @Override
+    public ServerStageVoiceChannel getChannel() {
+        return (ServerStageVoiceChannel) channel;
+    }
+
+}

--- a/javacord-core/src/main/java/org/javacord/core/event/channel/server/voice/ServerVoiceChannelEventImpl.java
+++ b/javacord-core/src/main/java/org/javacord/core/event/channel/server/voice/ServerVoiceChannelEventImpl.java
@@ -12,7 +12,7 @@ public abstract class ServerVoiceChannelEventImpl extends ServerEventImpl implem
     /**
      * The channel of the event.
      */
-    private final ServerVoiceChannel channel;
+    protected final ServerVoiceChannel channel;
 
     /**
      * Creates a new voice channel event.

--- a/javacord-core/src/main/java/org/javacord/core/util/handler/channel/ChannelCreateHandler.java
+++ b/javacord-core/src/main/java/org/javacord/core/util/handler/channel/ChannelCreateHandler.java
@@ -144,7 +144,7 @@ public class ChannelCreateHandler extends PacketHandler {
      */
     private void handlePrivateChannel(JsonNode channel) {
         // A CHANNEL_CREATE packet is sent every time a bot account receives a message, see
-        // https://github.com/hammerandchisel/discord-api-docs/issues/184Messa
+        // https://github.com/hammerandchisel/discord-api-docs/issues/184
 
         UserImpl recipient = new UserImpl(api, channel.get("recipients").get(0), (MemberImpl) null, null);
         if (!recipient.getPrivateChannel().isPresent()) {

--- a/javacord-core/src/main/java/org/javacord/core/util/handler/channel/ChannelCreateHandler.java
+++ b/javacord-core/src/main/java/org/javacord/core/util/handler/channel/ChannelCreateHandler.java
@@ -7,6 +7,7 @@ import org.javacord.api.entity.channel.ChannelCategory;
 import org.javacord.api.entity.channel.ChannelType;
 import org.javacord.api.entity.channel.GroupChannel;
 import org.javacord.api.entity.channel.PrivateChannel;
+import org.javacord.api.entity.channel.ServerStageVoiceChannel;
 import org.javacord.api.entity.channel.ServerTextChannel;
 import org.javacord.api.entity.channel.ServerVoiceChannel;
 import org.javacord.api.event.channel.group.GroupChannelCreateEvent;
@@ -52,6 +53,9 @@ public class ChannelCreateHandler extends PacketHandler {
                 break;
             case SERVER_VOICE_CHANNEL:
                 handleServerVoiceChannel(packet);
+                break;
+            case SERVER_STAGE_VOICE_CHANNEL:
+                handleServerStageVoiceChannel(packet);
                 break;
             case GROUP_CHANNEL:
                 handleGroupChannel(packet);
@@ -119,13 +123,28 @@ public class ChannelCreateHandler extends PacketHandler {
     }
 
     /**
+     * Handles server stage voice channel creation.
+     *
+     * @param channel The channel data.
+     */
+    private void handleServerStageVoiceChannel(JsonNode channel) {
+        long serverId = channel.get("guild_id").asLong();
+        api.getPossiblyUnreadyServerById(serverId).ifPresent(server -> {
+            ServerStageVoiceChannel voiceChannel = ((ServerImpl) server).getOrCreateServerStageVoiceChannel(channel);
+            ServerChannelCreateEvent event = new ServerChannelCreateEventImpl(voiceChannel);
+
+            api.getEventDispatcher().dispatchServerChannelCreateEvent((DispatchQueueSelector) server, server, event);
+        });
+    }
+
+    /**
      * Handles a private channel creation.
      *
      * @param channel The channel data.
      */
     private void handlePrivateChannel(JsonNode channel) {
         // A CHANNEL_CREATE packet is sent every time a bot account receives a message, see
-        // https://github.com/hammerandchisel/discord-api-docs/issues/184
+        // https://github.com/hammerandchisel/discord-api-docs/issues/184Messa
 
         UserImpl recipient = new UserImpl(api, channel.get("recipients").get(0), (MemberImpl) null, null);
         if (!recipient.getPrivateChannel().isPresent()) {

--- a/javacord-core/src/main/java/org/javacord/core/util/handler/channel/ChannelDeleteHandler.java
+++ b/javacord-core/src/main/java/org/javacord/core/util/handler/channel/ChannelDeleteHandler.java
@@ -8,6 +8,7 @@ import org.javacord.api.entity.channel.ChannelCategory;
 import org.javacord.api.entity.channel.ChannelType;
 import org.javacord.api.entity.channel.GroupChannel;
 import org.javacord.api.entity.channel.ServerChannel;
+import org.javacord.api.entity.channel.ServerStageVoiceChannel;
 import org.javacord.api.entity.channel.ServerTextChannel;
 import org.javacord.api.entity.channel.ServerVoiceChannel;
 import org.javacord.api.entity.channel.TextChannel;
@@ -50,6 +51,9 @@ public class ChannelDeleteHandler extends PacketHandler {
                 break;
             case SERVER_VOICE_CHANNEL:
                 handleServerVoiceChannel(packet);
+                break;
+            case SERVER_STAGE_VOICE_CHANNEL:
+                handleServerStageVoiceChannel(packet);
                 break;
             case GROUP_CHANNEL:
                 handleGroupChannel(packet);
@@ -122,6 +126,24 @@ public class ChannelDeleteHandler extends PacketHandler {
         api.getPossiblyUnreadyServerById(serverId)
                 .flatMap(server -> server.getVoiceChannelById(channelId))
                 .ifPresent(this::dispatchServerChannelDeleteEvent);
+        api.removeObjectListeners(ServerVoiceChannel.class, channelId);
+        api.removeObjectListeners(ServerChannel.class, channelId);
+        api.removeObjectListeners(VoiceChannel.class, channelId);
+        api.removeObjectListeners(Channel.class, channelId);
+    }
+
+    /**
+     * Handles server stage voice channel deletion.
+     *
+     * @param channelJson The channel data.
+     */
+    private void handleServerStageVoiceChannel(JsonNode channelJson) {
+        long serverId = channelJson.get("guild_id").asLong();
+        long channelId = channelJson.get("id").asLong();
+        api.getPossiblyUnreadyServerById(serverId)
+                .flatMap(server -> server.getStageVoiceChannelById(channelId))
+                .ifPresent(this::dispatchServerChannelDeleteEvent);
+        api.removeObjectListeners(ServerStageVoiceChannel.class, channelId);
         api.removeObjectListeners(ServerVoiceChannel.class, channelId);
         api.removeObjectListeners(ServerChannel.class, channelId);
         api.removeObjectListeners(VoiceChannel.class, channelId);

--- a/javacord-core/src/main/java/org/javacord/core/util/handler/channel/ChannelUpdateHandler.java
+++ b/javacord-core/src/main/java/org/javacord/core/util/handler/channel/ChannelUpdateHandler.java
@@ -18,11 +18,13 @@ import org.javacord.api.event.channel.server.ServerChannelChangeOverwrittenPermi
 import org.javacord.api.event.channel.server.ServerChannelChangePositionEvent;
 import org.javacord.api.event.channel.server.text.ServerTextChannelChangeSlowmodeEvent;
 import org.javacord.api.event.channel.server.text.ServerTextChannelChangeTopicEvent;
+import org.javacord.api.event.channel.server.voice.ServerStageVoiceChannelChangeTopicEvent;
 import org.javacord.api.event.channel.server.voice.ServerVoiceChannelChangeBitrateEvent;
 import org.javacord.api.event.channel.server.voice.ServerVoiceChannelChangeUserLimitEvent;
 import org.javacord.core.entity.channel.ChannelCategoryImpl;
 import org.javacord.core.entity.channel.GroupChannelImpl;
 import org.javacord.core.entity.channel.ServerChannelImpl;
+import org.javacord.core.entity.channel.ServerStageVoiceChannelImpl;
 import org.javacord.core.entity.channel.ServerTextChannelImpl;
 import org.javacord.core.entity.channel.ServerVoiceChannelImpl;
 import org.javacord.core.entity.permission.PermissionsImpl;
@@ -34,6 +36,7 @@ import org.javacord.core.event.channel.server.ServerChannelChangeOverwrittenPerm
 import org.javacord.core.event.channel.server.ServerChannelChangePositionEventImpl;
 import org.javacord.core.event.channel.server.text.ServerTextChannelChangeSlowmodeEventImpl;
 import org.javacord.core.event.channel.server.text.ServerTextChannelChangeTopicEventImpl;
+import org.javacord.core.event.channel.server.voice.ServerStageVoiceChannelChangeTopicEventImpl;
 import org.javacord.core.event.channel.server.voice.ServerVoiceChannelChangeBitrateEventImpl;
 import org.javacord.core.event.channel.server.voice.ServerVoiceChannelChangeUserLimitEventImpl;
 import org.javacord.core.util.event.DispatchQueueSelector;
@@ -81,6 +84,11 @@ public class ChannelUpdateHandler extends PacketHandler {
                 handleServerChannel(packet);
                 handleServerVoiceChannel(packet);
                 break;
+            case SERVER_STAGE_VOICE_CHANNEL:
+                handleServerChannel(packet);
+                handleServerVoiceChannel(packet);
+                handleServerStageVoiceChannel(packet);
+                break;
             case GROUP_CHANNEL:
                 handleGroupChannel(packet);
                 break;
@@ -89,7 +97,10 @@ public class ChannelUpdateHandler extends PacketHandler {
                 handleChannelCategory(packet);
                 break;
             case SERVER_NEWS_CHANNEL:
+                logger.debug("Received CHANNEL_UPDATE packet for a news channel. In this Javacord version it is "
+                        + "treated as a normal text channel!");
                 handleServerChannel(packet);
+                handleServerTextChannel(packet);
                 break;
             case SERVER_STORE_CHANNEL:
                 // TODO Handle store channels
@@ -343,9 +354,34 @@ public class ChannelUpdateHandler extends PacketHandler {
                         new ServerVoiceChannelChangeUserLimitEventImpl(channel, newUserLimit, oldUserLimit);
 
                 api.getEventDispatcher().dispatchServerVoiceChannelChangeUserLimitEvent(
-                        (DispatchQueueSelector) channel.getServer(), channel.getServer(), channel, event);
+                        (DispatchQueueSelector) channel.getServer(), channel.getServer(), channel, event
+                );
             }
         });
+    }
+
+    /**
+     * Handles a server stage voice channel update.
+     *
+     * @param jsonChannel The channel data.
+     */
+    private void handleServerStageVoiceChannel(JsonNode jsonChannel) {
+        long channelId = jsonChannel.get("id").asLong();
+        api.getServerStageVoiceChannelById(channelId)
+                .map(ServerStageVoiceChannelImpl.class::cast).ifPresent(channel -> {
+                    String oldTopic = channel.getTopic().orElse(null);
+                    String newTopic = jsonChannel.hasNonNull("topic")
+                            ? jsonChannel.get("topic").asText()
+                            : null;
+                    if (!Objects.equals(oldTopic, newTopic)) {
+                        channel.setTopic(newTopic);
+                        ServerStageVoiceChannelChangeTopicEvent event =
+                                new ServerStageVoiceChannelChangeTopicEventImpl(channel, newTopic, oldTopic);
+                        api.getEventDispatcher().dispatchServerStageVoiceChannelChangeTopicEvent(
+                                (DispatchQueueSelector) channel.getServer(), channel.getServer(), channel, event
+                        );
+                    }
+                });
     }
 
     /**

